### PR TITLE
add feature for region self to use ec2metadata service for configuration

### PR DESCRIPTION
**Pull request by trjstewart**
### Proposed changes
implements solution for issue #26 - use aws ec2 metadata for region config

### Checklist
Before creating a PR, run through this checklist and mark each as complete.

- [x] I have read the [CONTRIBUTING](https://github.com/nginxinc/nginx-asg-sync/blob/master/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that all unit tests pass after adding my changes
- [x] I have updated necessary documentation
- [x] I have rebased my branch onto master
- [x] I will ensure my PR is targeting the master branch and pulling from my branch from my own fork

### CODE REVIEW

**Comment by trjstewart** [comment]
I am currently unsure how I can add tests to cover the feature given the reliance on the ec2metadata service, open to suggestions.

File: cmd/sync/aws.go [hunk 1]
```diff
@@ -29,6 +30,24 @@ func NewAWSClient(data []byte) (*AWSClient, error) {
...
 	}
 
+	if cfg.Region == "self" {
+		metaSession, err := session.NewSession()
+		if err != nil {
+			return nil, err
+		}
+		
+		metaClient := ec2metadata.New(metaSession)
+		if metaClient.Available() != true {
```
Comment (pleshakov) [inline cmd/sync/aws.go]:
```suggestion
		if !metaClient.Available() {
```
for consistency with the rest of the code

File: cmd/sync/aws.go [hunk 2]
```diff
@@ -29,6 +30,24 @@ func NewAWSClient(data []byte) (*AWSClient, error) {
 		return nil, fmt.Errorf("error validating config: %v", err)
 	}
 
+	if cfg.Region == "self" {
+		metaSession, err := session.NewSession()
```
Comment (pleshakov) [inline cmd/sync/aws.go]: I suggest using the cfg with the same timeout value as in the https://github.com/nginxinc/nginx-asg-sync/blob/ae94a73cefcec28990c283904f6f13502ad15e36/cmd/sync/aws.go#L79

Comment (trjstewart) [reply cmd/sync/aws.go]: good suggestion. I didn't stick with the use of `cfg` due to it being consistently used in reference to the yaml config. I opted to use `params` instead.

**Comment by trjstewart** [comment]
> It is better to leave the region field so that we retain backward compatibility. I think the suggested special value self works OK. Alternatively, we could make the region field optional, so that if it is not set, then the asg-sync will try to discover the region.

A colleague of mine did actually raise this same question. Personally I am more a fan of declarative configuration rather than something that could be attributed to a typo or the like.

**Pull request merged**